src/hotspot/share/utilities/resourceHash.hpp

```diff
@@ -184,11 +184,20 @@ class ResourceHashtable : public ResourceObj {
   // the iteration is cancelled.
   template<class ITER>
   void iterate(ITER* iter) const {
+    auto function = [&] (K& k, V& v) {
```

Comment:
Can we avoid doing this by making a named `struct` implementing `do_entry` for every use? I realize that would deviate the _uses_ from the upstream, but we would not be changing the shared code, which would be safer for generic code that does not use any of async logging *and* uses older toolchains.

This hunk effectively backports https://bugs.openjdk.org/browse/JDK-8276789, which seems to imply the need to check this works well with different toolchains.

Reply:
yes, I can rewrite lambdas using old-style functor. 

to backport [JDK-8276789](https://bugs.openjdk.org/browse/JDK-8276789), we also need to backport bucket_at(), table()/table_size() beforehand. I take your advice. my approach skips more verification and it's not good.  I will use functor.